tests fail due to missing class filter AdRight2

For #64 the unit tests fail because the class based filter `.AdRight2` is not included in the resulting configuration:

```
 E           assert False
E            +  where False = check_not_in('AdRight2', '<html>\n\t<body>\n\t\t<div>just-some-test-string-always-present</div>\n\t\t\n\t\t<div class="asd ajlkl AdRight2">multiple classes that should be removed</div>\n\t\t<div class="asd ajlkl"> multiple classes that should exist </div>\n\t\t\n\t\t<div data-ad-manager-id class="success-valid-class">name-only attibute should be removed</div>\n\t\t\n\t\t\n\t\t\n\t\t\n\t\t<div class="adDisplay-modul">startswith match attribute should be exist</div>\n\t\t\n\t</body>\n</html>\n')
E            +    where '<html>\n\t<body>\n\t\t<div>just-some-test-string-always-present</div>\n\t\t\n\t\t<div class="asd ajlkl AdRight2">multiple classes that should be removed</div>\n\t\t<div class="asd ajlkl"> multiple classes that should exist </div>\n\t\t\n\t\t<div data-ad-manager-id class="success-valid-class">name-only attibute should be removed</div>\n\t\t\n\t\t\n\t\t\n\t\t\n\t\t<div class="adDisplay-modul">startswith match attribute should be exist</div>\n\t\t\n\t</body>\n</html>\n' = <Response [200]>.text
```

--- a/helper/install_deps.sh
+++ b/helper/install_deps.sh
@@ -50,8 +50,6 @@ certificate-directory /var/lib/privoxy/certs
 trusted-cas-file /etc/ssl/certs/ca-certificates.crt
 ca-cert-file cacert.crt
 ca-key-file cakey.pem
-# activate debugging of rules & access log
-debug 8704
 EOF
         fi
         if [ -e /usr/local/share/ca-certificates/privoxy/privoxy-cacert.crt ]; then
@@ -60,6 +58,12 @@ EOF
         ln -s /etc/privoxy/CA/cacert.crt /usr/local/share/ca-certificates/privoxy/privoxy-cacert.crt
         update-ca-certificates
         c_rehash
+    fi
+    if ! grep -q '^debug' /etc/privoxy/config; then
+        cat >> /etc/privoxy/config << EOF
+# activate debugging of rules & access log
+debug 8704
+EOF
     fi
     exit 0
 fi
@@ -116,6 +120,11 @@ EOF
         list    actionsfile             '/etc/config/privoxy_https.action'
 EOF
         fi
+    fi
+    if ! grep -q '^debug' /etc/config/privoxy; then
+        cat >> /etc/config/privoxy << EOF
+        option  debug             '8704'
+EOF
     fi
     exit 0
 fi

--- a/privoxy-blocklist.sh
+++ b/privoxy-blocklist.sh
@@ -414,7 +414,7 @@ function main() {
                         # remove all combinations with attribute matching
                         /^##\..*\[.*/d
                         # remove all matches with combinators
-                        /^##\..*[>+~ ].*/d
+                        /^##\..*[>+~ :].*/d
                         # cleanup
                         s/^##\.//g
                         # prepare regex merging
@@ -422,7 +422,7 @@ function main() {
                     ' "${html_file}" | while read -r line; do
                         # number of matches within one rule impacts runtime of each request to modify the content
                         if [ "${#lines[@]}" -lt 1000 ]; then
-                            lines+=("$line")
+                            lines+=("${line}")
                             continue
                         fi
                         # complexity of regex impacts runtime of each request to modify the content
@@ -519,7 +519,7 @@ function main() {
                     lines=()
                     # using while-loop as privoxy cannot handle more than 2000 or-connected strings within one regex
                     sed -e '
-                        # only process gloabl classes
+                        # only process gloabl attributes
                         /^##\[[^=][^=]*$/!d
                         # remove all matches with combinators
                         /^##.*[>+~ ].*/d
@@ -529,6 +529,8 @@ function main() {
                         s/^\[\([^=][^=]*\)\]/\1/g
                         # convert dots
                         s/\.\([^\.]\)/\\.\1/g
+                        # convert combined attribute name-only matches (e.g. ##[data-freestar-ad][id])
+                        s/\]\s*\[/.*/g
                         s/$/|/
                     ' "${html_file}" | sort -u | while read -r line; do
                         # number of matches within one rule impacts runtime of each request to modify the content

--- a/tests/Dockerfile_alpine
+++ b/tests/Dockerfile_alpine
@@ -1,13 +1,14 @@
 FROM		alpine:latest
 COPY		tests/requirements.txt /requirements.txt
 COPY		helper/install_deps.sh /install_deps.sh
+RUN		install -d -o root -g root /pytest_cache
 # hadolint ignore=DL3018
-RUN		apk add --no-cache --quiet build-base linux-headers py3-pip python3-dev sudo \
-		&& pip install --no-cache-dir -qr /requirements.txt \
-		&& rm -f /requirements.txt \
-		&& install -d -o root -g root /pytest_cache \
-		&& /install_deps.sh \
-		&& rm -f /install_deps.sh \
-		&& bash -c "for f in /etc/privoxy/*.new; do cp -p \$f \${f%.*};done"
+RUN		apk add --no-cache --quiet build-base linux-headers py3-pip python3-dev sudo
+RUN		pip install --no-cache-dir --break-system-packages -qr /requirements.txt \
+		&& rm -f /requirements.txt
+RUN		/install_deps.sh
+RUN		bash -c "for f in /etc/privoxy/*.new; do cp -p \$f \${f%.*};done"
+RUN		/install_deps.sh \
+		&& rm -f /install_deps.sh
 WORKDIR		/app
 ENTRYPOINT	["pytest", "-v", "-s", "-o", "cache_dir=/pytest_cache", "--color", "yes"]

--- a/tests/Dockerfile_ubuntu
+++ b/tests/Dockerfile_ubuntu
@@ -1,6 +1,7 @@
 FROM		ubuntu:latest
 COPY		tests/requirements.txt /requirements.txt
 COPY		helper/install_deps.sh /install_deps.sh
+RUN		install -d -o root -g root /pytest_cache
 # hadolint ignore=DL3018
 ENV		DEBIAN_FRONTEND=noninteractive
 ENV		HTTPS_SUPPORT=true
@@ -12,13 +13,12 @@ RUN		apt-get update \
 			python3-dev \
 			sudo \
 			vim \
-		&& pip install --no-cache-dir -qr /requirements.txt \
-		&& rm -f /requirements.txt \
-		&& install -d -o root -g root /pytest_cache \
-		&& /install_deps.sh \
-		&& rm -f /install_deps.sh \
 		&& apt-get clean \
 		&& apt-get -y autoremove \
 		&& rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+RUN		pip install --no-cache-dir -r /requirements.txt --break-system-packages \
+		&& rm -f /requirements.txt
+RUN		/install_deps.sh \
+		&& rm -f /install_deps.sh
 WORKDIR		/app
 ENTRYPOINT	["pytest", "-v", "-s", "-o", "cache_dir=/pytest_cache", "--color", "yes"]

--- a/tests/config.py
+++ b/tests/config.py
@@ -7,6 +7,7 @@ content_removed = [
     "MyAdsId3",  # id match
     "AdRight2",  # class match with element having multiple classes
     "data-ad-manager-id",  # attribute match
+    "data-freestar-ad id",  # combined attribute match
     'data-role="tile-ads-module"',  # attribute exact match
     'onclick="content.ad/"',  # attribute contain match
     'class="adDisplay-module_foobar"',  # attribute startswith match

--- a/tests/response.html
+++ b/tests/response.html
@@ -6,6 +6,7 @@
 		<div class="asd ajlkl"> multiple classes that should exist </div>
 		<div id="MyAdsId3">id should be removed</div>
 		<div data-ad-manager-id class="success-valid-class">name-only attibute should be removed</div>
+		<div data-freestar-ad id>combined name-only attibute should be removed</div>
 		<div data-role="tile-ads-module" class="success-valid-class">exact match attibute should be removed</div>
 		<div onclick="foo_content.ad/-asd">1. contain match attribute should be removed</div>
 		<div onclick="content.ad/">2. contain match attribute should be removed</div>

--- a/tests/run.sh
+++ b/tests/run.sh
@@ -6,29 +6,31 @@ SCRIPT_DIR="$(dirname "$(readlink -f "$0")")"
 GIT_DIR="$(dirname "${SCRIPT_DIR}")"
 os="ubuntu"
 interactive=0
+skip=0
 
 while getopts ":io:r" opt; do
     case "${opt}" in
         "i")
             interactive=1
-            shift
+            skip="$((skip + 1))"
             ;;
         "o")
             os="${OPTARG}"
-            shift 2
+            skip="$((skip + 2))"
             ;;
         "r")
             rebuild="true"
-            shift
+            skip="$((skip + 1))"
             ;;
         ":")
             echo "${TMPNAME}: -${OPTARG} requires an argument" >&2
             exit 1
             ;;
         *)  ;;
-
     esac
 done
+
+shift "${skip}"
 
 img_tag="privoxy-blocklist-test:${os}"
 dockerfile="${SCRIPT_DIR}/Dockerfile_${os}"
@@ -40,7 +42,7 @@ if ! [ -f "${dockerfile}" ]; then
 fi
 
 if [ "${rebuild}" = "true" ] || ! docker image ls -q "${img_tag}" | grep -vq '^$' || [ "$(($(date +%s) - $(date -d "$(docker image ls "${img_tag}" --format '{{print .CreatedAt}}' | cut -d' ' -f1,2)" +%s)))" -gt 14400 ]; then
-    echo "building docker image"
+    echo "building docker image for ${os}"
     cd "${GIT_DIR}"
     if docker image ls --format '{{ .Repository }}:{{ .Tag }}' | grep -q "${img_tag}"; then
         docker image rm "${img_tag}" > /dev/null || true
@@ -50,10 +52,13 @@ if [ "${rebuild}" = "true" ] || ! docker image ls -q "${img_tag}" | grep -vq '^$
     fi
     docker build -q -t "${img_tag}" -f "${dockerfile}" .
     cd -
+    echo
 fi
 
 if [ "${interactive}" -eq 0 ]; then
+    echo "running tests on ${os}"
     docker run --rm -w /app -v "${GIT_DIR}:/app" -v "${pytest_cache}:/pytest_cache" "${img_tag}" "${@:-./tests}"
 else
+    echo "interactive mode on ${os}"
     docker run -ti --rm -w /app -v "${GIT_DIR}:/app" -v "${pytest_cache}:/pytest_cache" --entrypoint /bin/bash "${img_tag}"
 fi

